In [1]:
import numpy as np
import pandas as pd
from sklearn.utils import shuffle

input_file = "emnist-digits-train.csv"
df_train = pd.read_csv(input_file, header=None)

In [2]:
data_train = df_train.to_numpy()
data_train

array([[8, 0, 0, ..., 0, 0, 0],
       [9, 0, 0, ..., 0, 0, 0],
       [6, 0, 0, ..., 0, 0, 0],
       ...,
       [1, 0, 0, ..., 0, 0, 0],
       [1, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [3]:
data_train.shape

(240000, 785)

In [4]:
y_train = data_train[:, 0]
y_train

array([8, 9, 6, ..., 1, 1, 0], dtype=int64)

In [5]:
X_train = data_train[:, 1:]
X_train

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

# ExtraTreeClassifier

In [21]:
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.ensemble import RandomForestClassifier

ext_clf = ExtraTreesClassifier(random_state=0)
ext_clf.fit(X_train, y_train)

ext_scores = np.mean(cross_val_score(ext_clf, X_train, y_train, cv=10, scoring="accuracy"))
past = ext_scores

for x in range(1, 1000):
    ext_clf = ExtraTreesClassifier(random_state=x)
    ext_clf.fit(X_train, y_train)
    
    ext_scores = np.mean(cross_val_score(ext_clf, X_train, y_train, cv=10, scoring="accuracy"))
    
    if ext_scores > past:
        past = ext_scores
        print(f"{x}번째 정확도 : {[past]}")

2번째 정확도 : [0.9829500000000001]
3번째 정확도 : [0.9830125000000001]
4번째 정확도 : [0.9830208333333333]
13번째 정확도 : [0.9830791666666666]
29번째 정확도 : [0.9830874999999999]
31번째 정확도 : [0.9831458333333332]
34번째 정확도 : [0.98315]


KeyboardInterrupt: 

2번째 정확도 : [0.9829500000000001]

3번째 정확도 : [0.9830125000000001]

4번째 정확도 : [0.9830208333333333]

13번째 정확도 : [0.9830791666666666]

29번째 정확도 : [0.9830874999999999]

31번째 정확도 : [0.9831458333333332]

34번째 정확도 : [0.98315]

너무 오래 걸려서 42번째에서 끊음

In [24]:
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.model_selection import cross_val_score

ext_clf1 = ExtraTreesClassifier(random_state=34)
ext_clf1.fit(X_train, y_train)

np.mean(cross_val_score(ext_clf1, X_train, y_train, cv=10, scoring="accuracy"))

0.98315

In [25]:
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.model_selection import cross_val_score

ext_clf2 = ExtraTreesClassifier(random_state=31)
ext_clf2.fit(X_train, y_train)

np.mean(cross_val_score(ext_clf2, X_train, y_train, cv=10, scoring="accuracy"))

0.9831458333333332

In [26]:
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.model_selection import cross_val_score

ext_clf3 = ExtraTreesClassifier(random_state=29)
ext_clf3.fit(X_train, y_train)

np.mean(cross_val_score(ext_clf3, X_train, y_train, cv=10, scoring="accuracy"))

0.9830874999999999

# Result!

In [30]:
from sklearn.ensemble import VotingClassifier

voting_clf = VotingClassifier(estimators=[('ext1', ext_clf1),('ext2', ext_clf2),('ext3', ext_clf3)], voting='hard')

In [32]:
np.mean(cross_val_score(voting_clf, X_train, y_train, cv=10, scoring="accuracy"))

0.9835875